haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs

```diff
@@ -137,16 +138,52 @@ analyseTypechecked' ghcEnv opts modSummary renSource tcSource (mod', mspan) = do
           renamedRefs = fromMaybe [] (refsFromRenamed ctx altMap <$> renSource)
       rels = fromMaybe [] (relationsFromRenamed ctx altMap <$> renSource)
       decls = map daDecl declsAlts
+      docDecls = generateDocDecls refs
```

Comment:
It turns out that the frontend is not a good place. It's necessary to deduplicate the doc/uri facts; otherwise the kythe verifier errors out saying that multiple facts are equal. Besides the verifier errors, I don't think it's good to generate many duplicate facts anyway.

Because the frontend sink callback is called once for each `XRef`, it's hard to deduplicate the generated doc/uri facts without adding some internal states to it, which I want to avoid.

I think I can move most of the logic to GhcApiSupport.hs from Ghc.hs though, which would be closer to the frontend. What do you think of that?

Re: properly indexing the base packages, I was told from Judah that you had done some work there and I saw your changes for adding build targets. But they are all gone now after we migrated to a newer GHC version. And I think it'll be quite a work to maintain that whenever we upgrade GHC.

Since this doc/uri facts link to the source code anyway (although it's in the hackage site), it wouldn't be too bad in terms of usability. Hackage even supports a sort of cross-referencing in their source documents!

Reply:
#### Where to dedup

Let's clarify what makes the backend a better place than the frontend if so.

Assuming N unique base decls getting referred in a module, with K total references (K >> N): As I see, the K reference targets need to be grouped (here via roundtripping a Map or Set) to know which are the N decls that need a doc/uri fact. Either in backend or frontend.

In the frontend it would be just as easy to group on (PkgModule, Name, Maybe Span) which should be pretty unique, and only then construct the stringy doc/uri. Well, it could be done on the backend as well, but why would one do it (doc/uri clearly is not coming from GHC backend). But maybe I miss something.

Then there are C modules, which is usually not a big number, and deduplicating among these is not such a burden.

#### Doing dedup in general

I agree that it makes sense to do module-level deduplication, but cross-module deduping should be left to a more generic proto-level deduper on the output stream. The benefit is that 1) it is more generic thus easier to support and 2) per-module entries can be streamed out independently, avoiding memory hogging (didn't measure, but gut feeling and was an intention in the code).

Reply:
It's tricky to do from the frontend as it is because its callback is passed one `XRef` one by one. See https://github.com/google/haskell-indexer/blob/master/haskell-indexer-pipeline-ghckythe/src/Language/Haskell/Indexer/Pipeline/GhcKythe.hs#L57. `withTypeChecked` calls `collect` for each `XRef`. To do deduplication there, it needs to collect all the `XRef`s again and dedup, which seems silly to me.

As for "why dedup?", my problem is more on the practical side. The Kythe verifier fails when there are multiple facts with the same v-name. I know the Kythe pipeline does its own deduplication, which I think is great. But I should dedup here myself too.

My change for moving most logic to GhcApiSupport is almost done. I'll upload the change soon.

Reply:
Well, I would argue doing a per-module (that is, per-XRef IIRC) deduping is good enough, and then `collect` could get an extra argument to some mutable bloom filter based on `Raw.Entry` hashes. That way there would be a final deduping over all modules.

I would be very surprised if other indexers did exact deduping vs just some best-effort thing, especially that latter has linear (?) access and scales better. Not that Haskell is the largest indexing burden, yet..

Reply:
Sadly, per-XRef-dedup wasn't enough. That's why I added this code. BTW now the dedup happens only once in GhcApiSupport.hs.

Reply:
Correction. By "I added this code", I means "I added dedup code in GhcApiSupport.hs as well as in Ghc.hs". But this is moot now, I guess.

Reply:
Thanks! Had a look, certainly more compact.

However, are we on the same page with the fact that adding a final approximate deduplication on the raw output entries would satisfy the verifier? (As the verified toy codes are tiny and won't exercise the full output cache).

If so, do you think per-module exact deduping + global approximate deduping is an alternative to cross-module deduping?

As mentioned, I'm concerned that forcing data from all modules for an exact cross-module deduping will hurt linearity, both conceptually and in terms of peak memory usage.

Reply:
Now the dedup is done while processing `generateDocUriDecls` rather than after the fact. I hope it addresses your concern about the memory usage and linearity.